### PR TITLE
Extract TelescopeController from SalsaTelescope

### PR DESCRIPTION
**Pull request by symmsaur**
We want to achieve two things:
- Shrinking SalsaTelescope which is getting big
- More intuitive updating of telescope controller state and tracking

To do this we take out the parts relating to the telescope controller from SalsaTelescope. The SalsaTelescope remains as a facade but can in the future be broken up further.

By using a long lived task to do all communication to with one telescope controller we prepare for using less large locks in the future. Updating the telescope continuously (e.g while tracking) is moved into this task. This allows other parts of the system to send commands and observe the effects on the telescope without having to continuously drive it.

### CODE REVIEW

File: backend/src/telescope_controller.rs [hunk 1]
```diff
@@ -0,0 +1,351 @@
+use chrono::{DateTime, Utc};
```
Comment (symmsaur) [inline backend/src/telescope_controller.rs:1]: Ordering in this file should be worked through

File: backend/src/telescope_controller.rs [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+fn spawn_controller_task(state: Arc<Mutex<TelescopeControllerState>>) -> JoinHandle<()> {
+    tokio::spawn(controller_task_function(state))
+}
+
+async fn controller_task_function(state: Arc<Mutex<TelescopeControllerState>>) {
+    // commanded_horizontal: Option<Direction>,
+    let mut connection_established = false;
+
+    loop {
+        // 10 Hz update freq
```
Comment (symmsaur) [inline backend/src/telescope_controller.rs:320]: Is this a good freq? Or go lower, 4 hz is fairly instant still?

File: backend/src/telescope_controller.rs [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+            state_guard.most_recent_error = send_command(&mut stream, TelescopeCommand::Stop).err();
+            state_guard.commanded_horizontal = None;
+            connection_established = true;
+        }
+
+        if state.lock().unwrap().should_restart {
+            state.lock().unwrap().most_recent_error =
+                send_command(&mut stream, TelescopeCommand::Restart).err();
+            connection_established = false;
+            sleep_until(Instant::now() + Duration::from_secs(10)).await;
```
Comment (symmsaur) [inline backend/src/telescope_controller.rs:343]: The old code waited for 10 secs so I guess we have to? Can't just wait to create a connection directly?

File: backend/src/salsa_telescope.rs [hunk 4]
```diff
@@ -1,120 +1,30 @@
 use crate::telescope::Telescope;
+use crate::telescope_controller::TelescopeController;
```
Comment (symmsaur) [inline backend/src/salsa_telescope.rs:2]: Tests need updating

File: backend/src/telescope_controller.rs [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+            "Telescope did not respond with current direction".to_string(),
+        )),
+    }
+}
+
+fn spawn_controller_task(state: Arc<Mutex<TelescopeControllerState>>) -> JoinHandle<()> {
+    tokio::spawn(controller_task_function(state))
+}
+
+async fn controller_task_function(state: Arc<Mutex<TelescopeControllerState>>) {
```
Comment (symmsaur) [inline backend/src/telescope_controller.rs:315]: We're continuously setting most_recent_error to the last encountered error. Not sure that is quite what we had before but maybe is ok? Could have a queue of errors but does that help anyone?

**Comment by symmsaur** [comment]
@centowen I think this looks quite promising, what do you think? If good I can get going on updating the tests and actually testing that it works for real.

**Comment by centowen** [comment]
Sorry for slow reply. Got a bit stuck in Elden Ring. :) I think this looks like a great improvement!

File: backend/src/telescope_controller.rs (hunk 3 above)
Comment (centowen) [reply backend/src/telescope_controller.rs:343]: One question here. If I understand this code correctly, then when we are waiting for the telescope to restart we will hold the lock on the mutex. That means that the HTTP requests from frontend will not be able to finish until restart is complete. This is probably ok, looking at the frontend code it should simply wait for the requests to finish.

In a future version we could add some kind of restarting state to indicate in the frontend that it is happening. But that is almost certainly scope creep if we do it know.

File: backend/src/telescope_controller.rs (hunk 5 above)
Comment (symmsaur) [reply backend/src/telescope_controller.rs]: Ok then we let it be like this for now at least

**Comment by symmsaur** [comment]
Now works with the fake telescope controller. Needs test with real telescope controller, then can be merged.

**Pull request merged**